ddev/ddev
# feat: disable `ddev config` in parent folders of a project, but allow it in subfolders, fixes #6783, fixes #5493

## The Issue

- #6783
- #5493 

## How This PR Solves The Issue

Automatically disable `ddev config` in:
- all project parent directories,
- home and its parent directories,
- global config directory `~/.ddev` and its nested directories,
- `ddev/ddev` repository root directory. (I myself accidentally ran `ddev config` here a few times.)

Allow to run `ddev config` in the project's subdirectory.

TODO:

- [x] Check if we can allow `ddev config` in the project subdirectory.

## Manual Testing Instructions

Check that you cannot run `ddev config` in any parent of any existing project:
```
mkdir -p ~/workspace/directory/projects/my-project
cd ~/workspace/directory/projects/my-project && ddev config --auto

cd ..
ddev config
Could not create a new config: 'ddev config' is not recommended in the /home/stas/workspace/directory/projects directory because a project exists in the subdirectory /home/stas/workspace/directory/projects/my-project
To use 'ddev config' here, run 'ddev stop --unlist' for all projects in subdirectories of this directory first

cd ..
ddev config
Could not create a new config: 'ddev config' is not recommended in the /home/stas/workspace/directory directory because a project exists in the subdirectory /home/stas/workspace/directory/projects/my-project
To use 'ddev config' here, run 'ddev stop --unlist' for all projects in subdirectories of this directory first

cd ..
ddev config
Could not create a new config: 'ddev config' is not recommended in the /home/stas/workspace directory because a project exists in the subdirectory /home/stas/workspace/directory/projects/my-project
To use 'ddev config' here, run 'ddev stop --unlist' for all projects in subdirectories of this directory first
```

You cannot run it in your home or parent directories of your home:
```
cd ~
ddev config
Could not create a new config: 'ddev config' is not recommended in your home directory (/home/stas)

cd /home
ddev config
Could not create a new config: 'ddev config' is not recommended in the parent directories of your home directory (/home)

cd /
ddev config
Could not create a new config: 'ddev config' is not recommended in the parent directories of your home directory (/)
```

You cannot run it in your global config directory and its nested directories:
```
cd ~/.ddev
ddev config
Could not create a new config: 'ddev config' is not recommended in your global config directory (/home/stas/.ddev)

cd ~/.ddev/commands
ddev config
Could not create a new config: 'ddev config' is not recommended in your global config directory (/home/stas/.ddev/commands)
```

You cannot run it in the directory where you're developing DDEV itself:
```
cd ~/workspace/ddev
ddev config
Could not create a new config: 'ddev config' is not allowed in the directory used for DDEV development (/home/stas/workspace/ddev)
```

Allow `ddev config` in the project subdirectory:

```
mkdir -p ~/workspace/directory/projects/my-project
cd ~/workspace/directory/projects/my-project && ddev config --auto

cd .ddev

# before
ddev config --php-version=8.2
it usually does not make sense to `ddev config` in a subdirectory of an existing project. Is it possible you wanted to `ddev config` in parent directory /home/stas/workspace/directory/projects/my-project?

# after
ddev config --php-version=8.2
You are reconfiguring the project at /home/stas/workspace/directory/projects/my-project.
The existing configuration will be updated and replaced.
```

## Automated Testing Overview



## Release/Deployment Notes

This can break the user's workflow if they have a DDEV project inside another DDEV project, which I don't think is a good practice.
